processing/src/main/java/io/druid/segment/DimensionIndexer.java

```diff
@@ -0,0 +1,350 @@
+ * such as conversion between an encoded value and its full representation. It maintains knowledge of the
+ * mappings between encoded values and actual values.
+ *
+ *
+ * Sorting and Ordering
+ * --------------------
+ * When encoding is present, there are two relevant orderings for the encoded values.
+ *
+ * 1.) Ordering based on encoded value's order of ingestion
+ * 2.) Ordering based on converted actual value
+ *
+ * Suppose we have a new String dimension DimA, which sees the values "Hello", "World", and "Apple", in that order.
+ * This would correspond to dictionary encodings of "Hello"=0, "World"=1, and "Apple"=2, by the order
+ * in which these values were first seen during ingestion.
+ *
+ * However, some use cases require the encodings to be sorted by their associated actual values.
+ * In this example, that ordering would be "Apple"=0, "Hello"=1, "World"=2.
+ *
+ * The first ordering will be referred to as "Unsorted" in the documentation for this interface, and
+ * the second ordering will be referred to as "Sorted".
+ *
+ * The unsorted ordering is used during ingestion, within the IncrementalIndex's TimeAndDims keys; the encodings
+ * are built as rows are ingested, taking the order in which new dimension values are seen.
+ *
+ * The generation of a sorted encoding takes place during segment creation when indexes are merged/persisted.
+ * The sorted ordering will be used for dimension value arrays in that context and when reading from
+ * persisted segments.
+ *
+ * Note that after calling the methods below that deal with sorted encodings,
+ * - getSortedEncodedValueFromUnsorted()
+ * - getUnsortedEncodedValueFromSorted()
+ * - getSortedIndexedValues()
+ * - convertUnsortedEncodedArrayToSortedEncodedArray()
+ *
+ * calling processRowValsToUnsortedEncodedArray() afterwards can invalidate previously read sorted encoding values
+ * (i.e., new values could be added that are inserted between existing values in the ordering).
+ *
+ *
+ * Thread Safety
+ * --------------------
+ * Each DimensionIndexer exists within the context of a single IncrementalIndex. Before IndexMerger.persist() is
+ * called on an IncrementalIndex, any associated DimensionIndexers should be thread-safe.
+ *
+ * In current usage, this means that an implementation should allow multiple threads to add data to the indexer via
+ * processRowValsToUnsortedEncodedArray() and allow multiple threads to read data via methods that only deal with
+ * unsorted encodings.
+ *
+ * As mentioned in the "Sorting and Ordering" section, writes and calls to the sorted encoding
+ * methods should not be interleaved: the sorted encoding methods should only be called when it is known that
+ * writes to the indexer will no longer occur.
+ *
+ * Index merge and persist are currently handled by a single thread.
+ *
+ * After IndexMerger.persist() is called on an IncrementalIndex, any indexers associated with that IncrementalIndex
```

Comment:
In the current implementation, the Runnable that runs IndexMerger.persist() is submitted to an Executor by the same thread that handled the adds. This is the case in IndexGeneratorJob and the Plumbers, so that should take care of memory consistency between the threads according to this page:

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#MemoryVisibility

"Actions in a thread prior to the submission of a Runnable to an Executor happen-before its execution begins. Similarly for Callables submitted to an ExecutorService."

This patch keeps the current behavior at these thread boundaries